### PR TITLE
Added Longest Increasing Subsequence

### PR DESCRIPTION
**Pull request by tanyarajhans**

### CODE REVIEW

**Comment by czgdp1807** [comment]
From now onwards, keep updating this PR for longest increasing sub-sequence.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 1]
```diff
@@ -787,3 +717,42 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+    ==========
+    .. [1] https://en.wikipedia.org/wiki/Longest_increasing_subsequence
+
+    """
+    n=len(seq)
+    lis=[1]*n
+    for i in range(1, n):
+        for j in range(0, i):
+            if seq[i] > seq[j] and lis[i]< lis[j] + 1:
+                lis[i] = lis[j]+1
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:754]: It is still O(N^2). Please implement O(N log N) version of this algorithm. Update in this PR only. Do not make new PRs.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 2]
```diff
@@ -493,12 +449,9 @@ def bucket_sort(array: Array, **kwargs) -> Array:
 
     References
     ==========
-
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:496]: Please restore these lines.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 1 above)
Comment (tanyarajhans) [reply pydatastructs/linear_data_structures/algorithms.py:754]: Sure

File: pydatastructs/linear_data_structures/algorithms.py (hunk 2 above)
Comment (tanyarajhans) [reply pydatastructs/linear_data_structures/algorithms.py]: Done

**Comment by tanyarajhans** [comment]
Have added nlogn approach for LIS and restored the lines as mentioned before.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 3]
```diff
@@ -787,3 +788,65 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
                     check_mat[i][j] = check_mat[i][j-1]
 
     return OneDimensionalArray(seq1._dtype, check_mat[row][col][-1])
+
+def longest_increasing_subsequence(seq: OneDimensionalArray) -> int:
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:792]: The function currently returns only the length. There should be a possibility of obtaining the longest increasing sub-sequence as well.

For example, if `ODA([1, 2, -1, 3, 4, 5])` is the input then, `(ODA[1, 2, 3, 4, 5], 5)` should be the output.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 4]
```diff
@@ -787,3 +788,65 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+                r = m
+            else:
+                l = m
+        return r
+
+    size = len(seq)
+    tailTable = [0 for i in range(size + 1)]
+    length = 0 # always points empty slot
+
+    tailTable[0] = seq[0]
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:833]: Use [snake case](https://en.wikipedia.org/wiki/Snake_case) and not [camel case](https://en.wikipedia.org/wiki/Camel_case).

File: pydatastructs/utils/tests/test_code_quality.py [hunk 5]
```diff
@@ -22,7 +22,7 @@ def test_trailing_white_spaces():
             if line.endswith(" \n") or line.endswith("\t\n") \
                 or line.endswith(" ") or line.endswith("\t"):
                 assert False, "%s contains trailing whitespace at %s"\
-                               %(file_path, line)
+                              %(file_path, line)
```
Comment (czgdp1807) [inline pydatastructs/utils/tests/test_code_quality.py:25]:
```suggestion
                             %(file_path, line)
```

**Comment by czgdp1807** [comment]
Are you editing directly on Github?

**Comment by tanyarajhans** [comment]
> Are you editing directly on Github?

yeah, just for the last few commits, as CLI wasn't working properly. Will do it from CLI from tomorrow.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 3 above)
Comment (Smit-create) [reply pydatastructs/linear_data_structures/algorithms.py:792]: Please refer to this for restoring the sequence: https://cp-algorithms.com/sequences/longest_increasing_subsequence.html#toc-tgt-8

**Comment by Smit-create** [comment]
> This contribution is made as a part of my participation in GirlScript Summer of Code 2021.

Please provide some updates on the PR or if any difficulty is faced by you so that it can be resolved. As this was stalled for few weeks

File: pydatastructs/linear_data_structures/algorithms.py [hunk 6]
```diff
@@ -787,3 +788,65 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
...
+
+    """
+    def CeilIndex(A, l, r, key):
+        while (r - l > 1):
+            m = l + (r - l)//2
+            if (A[m] >= key):
+                r = m
+            else:
+                l = m
+        return r
```
[the remark this answers is not in the document]
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:827]: > Can we just define and use methods or functions on sorted ODA for finding lower_bound and upper_bound (similar to how it works in cpp)?

Sure.

**Comment by Smit-create** [comment]
ping @tanyarajhans

**Comment by tanyarajhans** [comment]
> ping @tanyarajhans

Hey I am really sorry, for some reason I just received the mail right now. I had exams going on. Will add printing LIS in nlogn in few days, is that fine?

**Comment by tanyarajhans** [comment]
Also, could you please assign this issue to me, as I can see no one is assigned.

**Comment by Smit-create** [comment]
You don't need to worry about the points. As when you will succeed in merging the PR, you will get the assigned points.

**Comment by Smit-create** [comment]
@tanyarajhans Why did you close it? Are you not willing to work on this further?

**Comment by sHiVaNgI821** [comment]
I would like to work on this issue as part of my contribution for GSSoC'21. This method would be a part of the algorithms.py file. I would also like to make use of the upper_bound function which I have recently raised a PR for. The time complexity for this method, with the use of the upper_bound function would be O(nlogn).

The function definition goes as follows:-
```
def longest_increasing_subsequence(array: OneDimensionalArray ) -> OneDimensionalArray : 
        # algo (returns a OneDimensionalArray, which is the longest increasing subsequence
                     for that array)
        return longestIncreasingSubsequence
```

**Comment by Smit-create** [comment]
Closing in favor of #363

**Pull request closed without merging**